cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -189,20 +201,26 @@ Status ReplaceWithArrayMask(KernelContext* ctx, const ArrayData& array,
     const ArrayData& array_repl = *replacements.array();
     ReplaceWithArrayMaskImpl<Functor>(
         array, mask, array_repl, replacements_bitmap,
-        CopyArrayBitmap{replacements_bitmap ? array_repl.buffers[0]->data() : nullptr,
-                        array_repl.offset},
+        CopyArrayBitmap{
+            (!!array_repl.buffers[0]) ? array_repl.buffers[0]->data() : nullptr,
+            array_repl.offset},
         mask_bitmap, mask_values, out_bitmap, out_values, out_offset);
   } else {
     const Scalar& scalar_repl = *replacements.scalar();
     ReplaceWithArrayMaskImpl<Functor>(array, mask, scalar_repl, replacements_bitmap,
-                                      CopyScalarBitmap{scalar_repl.is_valid}, mask_bitmap,
-                                      mask_values, out_bitmap, out_values, out_offset);
+                                      CopyScalarBitmap{scalar_repl.is_valid, nullptr},
+                                      mask_bitmap, mask_values, out_bitmap, out_values,
+                                      out_offset);
   }
 
   if (mask.MayHaveNulls()) {
     arrow::internal::BitmapAnd(out_bitmap, out_offset, mask.buffers[0]->data(),
                                mask.offset, array.length, out_offset, out_bitmap);
   }
+  if (!output->GetNullCount()) {
+    output->null_count = 0;
+    output->buffers[0] = NULLPTR;
```

Comment:
I think with preallocated bitmaps like this, we should not delete the output buffer like this (it breaks `can_write_into_slices`).

Reply:
(You can see this on line 503 - the kernel specifies the output validity buffer can be preallocated.)